### ISSUE TITLE
`tf.linalg.sqrtm` is slow for large matrices

### ISSUE THREAD

**Issue by kyamagu** [tensorflow/tensorflow#58334]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Performance

### Source

binary

### Tensorflow Version

2.9.2

### Custom Code

No

### OS Platform and Distribution

Linux 9ded5992957c 5.10.133+ x86_64 GNU/Linux (Google Colab)

### Python version

3.7.15

### Current Behaviour?

```shell
tf.linalg.sqrtm is slow compared to scipy.linalg.sqrtm for large matrices.

For matrices of 512x512, scipy takes 4.2s while tf takes 40.1s on CPU.
```

### Standalone code to reproduce the issue

```shell
import timeit

import matplotlib.pyplot as plt
import scipy.linalg
import tensorflow as tf

print("tensorflow version = {}".format(tf.__version__))

sizes = 2 ** tf.range(1, 10)  # Matrix size.
functions = [tf.linalg.sqrtm, scipy.linalg.sqrtm]  # TF and scipy impl.

# Measure time.
result = dict()
for size in sizes:
    X = tf.eye(size, dtype=tf.float64) + 1e-2 * tf.random.normal((size, size), dtype=tf.float64)
    X = tf.cast(X, tf.complex128)
    for func in functions:
        elapsed = timeit.timeit(lambda: func(X), number=3)
        result.setdefault(func.__module__, []).append(elapsed)

# Plot results.
for key in result:
    plt.plot(sizes, result[key])
plt.legend([key.split(".")[0] for key in result.keys()])
plt.title("execution time")
plt.ylabel("seconds")
plt.xlabel("matrix size")
plt.grid(True)
plt.show()
```

### Relevant log output

```shell
tensorflow version = 2.9.2

See notebook https://gist.github.com/kyamagu/e301b61f9931da0196afe457b4f87023
```
</details>

**Comment by kyamagu** [comment]
Here is the plot.
![download](https://user-images.githubusercontent.com/1190780/198203474-e91f9923-5080-4b0a-8fde-afb9b174f0d0.png)

**Comment by tilakrayal** [comment]
@sachinprasadhs, 
I was able to reproduce the issue on tensorflow v2.9, v2.10 and nightly. Kindly find the gist of it [here](https://colab.research.google.com/gist/SuryanarayanaY/10e9544a2b9153b2bb7f657c8806efb0/58334.ipynb).

**Comment by bhack** [comment]
https://github.com/JuliaLang/julia/issues/40207

https://jax.readthedocs.io/en/latest/_modules/jax/_src/scipy/linalg.html#sqrtm

**Comment by cantonios** [comment]
TF uses the same algorithm as JAX, so not the recursive blocking method.  The paper reference is quoted in the TF docs for [`tf.linalg.sqrtm`](https://www.tensorflow.org/api_docs/python/tf/linalg/sqrtm).

**Comment by cantonios** [comment]
TF just calls Eigen's matrix sqrt [here](https://github.com/tensorflow/tensorflow/blob/38a677b555140410598603ff15abbe408b252e40/tensorflow/core/kernels/linalg/matrix_square_root_op.cc#L45).

I don't think we have that in Eigen either.

**Comment by bhack** [comment]
> TF just calls Eigen's matrix sqrt [here](https://github.com/tensorflow/tensorflow/blob/38a677b555140410598603ff15abbe408b252e40/tensorflow/core/kernels/linalg/matrix_square_root_op.cc#L45).
> 
> I don't think we have that in Eigen either.

So, probably still waiting on https://gitlab.com/libeigen/eigen/-/issues/840

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.

**Comment by fegemo** [comment]
The performance issue is still present as of Tensorflow 2.18.0. Re-running the code from the original poster, I get:

![Image](https://github.com/user-attachments/assets/2c47de11-ba73-4ac9-b601-34b5454713b6)

**Comment by cantonios** [comment]
> The performance issue is still present as of Tensorflow 2.18.0. Re-running the code from the original poster, I get:
> 
> ![Image](https://github.com/user-attachments/assets/2c47de11-ba73-4ac9-b601-34b5454713b6)
> 
> Maybe reopening this issue but tagging it with the current versions would be the way to go? If not, I could open a new ticket.

We have no planned action on this.  Someone can open a PR in Eigen for a blocked version of `sqrtm`, but there's nothing to do on the TensorFlow side.
